dask_glm/regularizers.py

```diff
@@ -3,85 +3,123 @@
+        raise NotImplementedError
 
-    @staticmethod
-    def add_reg_f(f, lam):
+    def add_reg_f(self, f, lam):
+        """Add regularization function to other function."""
         def wrapped(beta, *args):
-            return f(beta, *args) + lam * L2.f(beta)
+            return f(beta, *args) + lam * self.f(beta)
         return wrapped
 
-    @staticmethod
-    def gradient(beta):
-        return 2 * beta
+    def add_reg_grad(self, grad, lam):
+        """Add regularization gradient to other gradient function."""
+        def wrapped(beta, *args):
+            return grad(beta, *args) + lam * self.gradient(beta)
+        return wrapped
 
-    @staticmethod
-    def add_reg_grad(grad, lam):
+    def add_reg_hessian(self, hess, lam):
+        """Add regularization hessian to other hessian function."""
         def wrapped(beta, *args):
-            return grad(beta, *args) + lam * L2.gradient(beta)
+            return hess(beta, *args) + lam * self.hessian(beta)
         return wrapped
 
+    @classmethod
+    def get(cls, obj):
+        if isinstance(obj, cls):
+            return obj
+        elif isinstance(obj, str):
+            return {o._name: o for o in cls.__subclasses__()}[obj]()
+        raise TypeError('Not a valid regularizer object.')
 
-class L1(object):
 
-    @staticmethod
-    def proximal_operator(beta, t):
-        z = np.maximum(0, beta - t) - np.maximum(0, -beta - t)
-        return z
+class L2(Regularizer):
+    """L2 regularization."""
+    _name = 'l2'
 
-    @staticmethod
-    def hessian(beta):
-        raise ValueError('l1 norm is not twice differentiable!')
+    def f(self, beta):
+        return (beta**2).sum() / 2
 
-    @staticmethod
-    def add_reg_hessian(hess, lam):
-        def wrapped(beta, *args):
-            return hess(beta, *args) + lam * L1.hessian(beta)
-        return wrapped
+    def gradient(self, beta):
+        return beta
 
-    @staticmethod
-    def f(beta):
-        return (np.abs(beta)).sum()
+    def hessian(self, beta):
+        return np.eye(len(beta))
+
+    def proximal_operator(self, beta, t):
+        return 1 / (1 + t) * beta
 
-    @staticmethod
-    def add_reg_f(f, lam):
-        def wrapped(beta, *args):
-            return f(beta, *args) + lam * L1.f(beta)
-        return wrapped
 
-    @staticmethod
-    def gradient(beta):
+class L1(Regularizer):
+    """L1 regularization."""
+    _name = 'l1'
+
+    def f(self, beta):
+        return (np.abs(beta)).sum()
+
+    def gradient(self, beta):
         if np.any(np.isclose(beta, 0)):
             raise ValueError('l1 norm is not differentiable at 0!')
         else:
             return np.sign(beta)
 
-    @staticmethod
-    def add_reg_grad(grad, lam):
-        def wrapped(beta, *args):
-            return grad(beta, *args) + lam * L1.gradient(beta)
-        return wrapped
+    def hessian(self, beta):
+        raise ValueError('l1 norm is not twice differentiable!')
```

Comment:
Then do we want to switch the elastic net hessian to include the l1 side to the weight?  It won't have an effect except raise when there are errors.